stdlib/public/Atomics/AtomicProtocols.swift

```diff
@@ -0,0 +1,247 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public protocol AtomicProtocol {
+  associatedtype AtomicStorage
+
+  static func atomicStorage(for value: Self) -> AtomicStorage
+
+  static func deinitializeAtomicStorage(
+    at address: UnsafeMutablePointer<AtomicStorage>
+  )
+
+  @_semantics("has_constant_evaluable_arguments")
+  static func atomicLoad(
+    at address: UnsafeMutablePointer<AtomicStorage>,
+    ordering: AtomicLoadOrdering
+  ) -> Self
+
+  @_semantics("has_constant_evaluable_arguments")
+  static func atomicStore(
+    at address: UnsafeMutablePointer<AtomicStorage>,
+    _ desired: Self,
+    ordering: AtomicStoreOrdering
+  )
+
+  @_semantics("has_constant_evaluable_arguments")
+  static func atomicExchange(
+    at address: UnsafeMutablePointer<AtomicStorage>,
+    _ desired: Self,
+    ordering: AtomicUpdateOrdering
+  ) -> Self
+
+  @_semantics("has_constant_evaluable_arguments")
+  static func atomicCompareExchange(
+    at address: UnsafeMutablePointer<AtomicStorage>,
+    expected: Self,
+    desired: Self,
+    ordering: AtomicUpdateOrdering
+  ) -> (exchanged: Bool, original: Self)
+
+  @_semantics("has_constant_evaluable_arguments")
+  static func atomicCompareExchange(
+    at address: UnsafeMutablePointer<AtomicStorage>,
+    expected: Self,
+    desired: Self,
+    ordering: AtomicUpdateOrdering, // Note: no default
+    failureOrdering: AtomicLoadOrdering // Note: no default
+  ) -> (exchanged: Bool, original: Self)
+}
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension AtomicProtocol where AtomicStorage == Self {
+  @inlinable
+  public static func atomicStorage(for value: Self) -> AtomicStorage {
+    return value
+  }
+
+  @inlinable
+  public static func deinitializeAtomicStorage(
+    at address: UnsafeMutablePointer<AtomicStorage>
+  ) {
+    address.deinitialize(count: 1)
+  }
+}
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public protocol AtomicInteger: AtomicProtocol, FixedWidthInteger {
```

Comment:
No atomic increment/decrement for pointers? (Could be useful for lockless data structures.)

Reply:
Hm -- `AtomicAdditiveArithmetic`?

Reply:
AtomicStrideable, I think. That's what I put in my prototype, and it can't be AdditiveArithmetic because the types don't match up.

Reply:
Sounds good! So we'll have this tree of protocols:

```
AtomicProtocol → AtomicStrideable → AtomicInteger
                           → NullableAtomic
```

Reply:
Except not at all -- the stride needs to be independent of `Self` :-P

Reply:
Well, I guess `AtomicInteger` could constrain `Stride == Self`. 🤔

Reply:
:-/ In my API prototype I had the backing storage of pointers be Int for this reason. The strideable-ness is also still important because the unit stride of, e.g. `UnsafePointer<Int32>` is 4, not 1.

Reply:
Hm -- I keep flip-flopping on backing pointers/`Unmanaged` with `Int`, but implementing arithmetic doesn't seem like a huge issue either way.

We need to be able to model integer types narrower (and probably wider) than `Int` though.

Reply:
\*shrug\* Requiring that `Stride` be the same as `Self` seems reasonable for AtomicInteger, but you could also just make it a default implementation for the map-Stride-into-AtomicStorage operation on AtomicStrideable (if you go with my route).

Reply:
Let's see where the implementation takes us. 

Hm, are we to define pointer arithmetic for `Optional<UnsafePointer>`, too?

Reply:
Here was my unhappy best attempt:

```
// Optional does not (and should not) conform to Strideable...
// ...but it seems useful to be able to move pointers around atomically...
// ...but it's not type-safe when the value is nil.
//
// Actually, more specifically: loadThenIncrement can trap on nil, but incrementThenLoad can't.
//
// So we could do something like this (using Alejandro Alonso's work on parameterized extensions, https://forums.swift.org/t/parameterized-extensions/25563).

#if compiler(>=6)
extension <Wrapped> UnsafeAtomicPointer where
    Value == Optional<Wrapped>,
    Wrapped: AtomicStrideable,
    Wrapped: AtomicOptionalRepresentable {
  // The two meanings of "wrapped" here is unfortunate.
  func loadUnwrappedThenWrappingIncrement(
    by operand: Wrapped.Stride,
    ordering: AtomicUpdateOrdering
  ) -> Wrapped {
    let rawStride = Wrapped.atomicStorageRepresentation(of: operand)
    let rawResult = address.withMemoryRebound(to: Value.AtomicStorage.self, capacity: 1) {
      Value.AtomicStorage.atomicLoadThenWrappingIncrement(from: $0, by: rawStride, ordering: ordering)
    }
    // Note that we're still checking if the loaded value is nil.
    return unsafeBitCast(result, to: Optional<Wrapped>.self)!
  }
}
#endif
```

Reply:
I think we mustn't allow the results of nil pointer arithmetic to get committed into memory -- it would allow other threads to go astray before the trap happens. (So the `loadThen` and the `ThenLoad` forms must always have the same observable effect.)

Parameterized extensions can be replaced by more protocols; but is this really something that we need to model?

Reply:
Oof, I didn't think about other threads doing work before the trap. Maybe there's no safe way to do this. It's really too bad, though, because a compare-and-swap-based update will be needlessly slower.

(We could also say "it's undefined behavior to increment a nil pointer" but even for an unsafe API that's pretty bad.)

Reply:
Wait -- what's the envisioned use case for atomic pointer arithmetic again?

The pointer will necessarily point into some bounded buffer, and a direct atomic +1 won't be able to do any bounds checking before incrementing the value -- so if there is a high enough number of competing threads, they could cause the pointer to escape its buffer.

Reply:
You could do a fast atomic increment and then check for validity afterward, but yeah, maybe I'm just hypothesizing things people don't actually do. Maybe someone else on the thread can be called in to check?

Reply:
That runs into the same problem as null pointer arithmetic above -- we must not allow invalid state to get atomically committed into memory. It sounds like pointer arithmetic will need to be always done with explicit compare/exchange loops. (Unless CPU vendors vastly expand their atomic instruction sets...)

Reply:
Yeah, oh well. Maybe that means AtomicStrideable's not worth it. (On the other hand, if it turns out to be useful, we don't yet have a way to add it in a later release…)

Reply:
I'm pretty sure we'll be able to eventually add it if CPUs get bounded atomic increments -- it's "just" a matter of adding a new protocol or two.